### PR TITLE
SOLR-13941: Configure JettySolrRunner same as in web.xml

### PR DESCRIPTION
**Pull request by janhoy**
# Description

Make sure tests have same servlet root for DispatchFilter as when running Solr from cmdline. This avoids some subtle confusion in tests.

# Solution

Wire SolrDispatchFilter on `/*` instead of `*`, and make a new `SerlvetUtils` class with a method that concatenates servletPath and pathInfo for more readable and less error prone code.

# Tests

No new tests necessary

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

**Comment by uschindler** [comment]
You can remove the dummy 404Servlet from source code. It's obsolete now.

**Comment by janhoy** [comment]
Ok, @uschindler I removed the 404 servlet and fixed a test. I believe there may be a test or two that fails consistently still, here are the three fails from my last run

    [junit4]   - org.apache.solr.cloud.TestTlogReplica.testRemoveLeader
    [junit4]   - org.apache.solr.update.processor.DimensionalRoutedAliasUpdateProcessorTest.testCatTime
    [junit4]   - org.apache.solr.handler.admin.CoreAdminHandlerTest.testUnloadForever

**Comment by janhoy** [comment]
New test run has these fails

    [junit4] Tests with failures [seed: 27EC334A32BB0961]:
    [junit4]   - org.apache.solr.cloud.TestAuthenticationFramework.testBasics
    [junit4]   - org.apache.solr.cloud.TestTlogReplica.testRemoveLeader

Which suggests there may be an issue with `TestTlogReplica.testRemoveLeader` for this patch?

Stack trace:

    [junit4]   2> NOTE: reproduce with: ant test  -Dtestcase=TestTlogReplica -Dtests.method=testRemoveLeader -Dtests.seed=27EC334A32BB0961 -Dtests.slow=true  -Dtests.badapples=true -Dtests.locale=sr-Latn-XK -Dtests.timezone=Europe/Isle_of_Man -Dtests.asserts=true -Dtests.file.encoding=ISO-8859-1
    [junit4] ERROR   8.10s J1 | TestTlogReplica.testRemoveLeader <<<
    [junit4]    > Throwable #1: org.apache.solr.client.solrj.impl.CloudSolrClient$RouteException: Error from server at http://127.0.0.1:51091/solr/ tlog_replica_test_remove_leader_shard1_replica_t2: Expected mime type application/octet-stream but got text/html. <html>
    [junit4]    > <head>
    [junit4]    > <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
    [junit4]    > <title>Error 405 HTTP method POST is not supported by this URL</title>
    [junit4]    > </head>
    [junit4]    > <body><h2>HTTP ERROR 405</h2>
    [junit4]    > <p>Problem accessing /solr/tlog_replica_test_remove_leader_shard1_replica_t2/update. Reason:
    [junit4]    > <pre>    HTTP method POST is not supported by this URL</pre></p><hr><a href="http://eclipse.org/jetty">Powered by Jetty://  9.4.19.v20190610</a><hr/>
    [junit4]    > </body>
    [junit4]    > </html>
    [junit4]    > 	at __randomizedtesting.SeedInfo.seed([27EC334A32BB0961:BEACF42E7532692B]:0)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.CloudSolrClient.getRouteException(CloudSolrClient.java:125)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.CloudSolrClient.getRouteException(CloudSolrClient.java:46)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.BaseCloudSolrClient.directUpdate(BaseCloudSolrClient.java:549)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.BaseCloudSolrClient.sendRequest(BaseCloudSolrClient.java:1036)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.BaseCloudSolrClient.requestWithRetryOnStaleState(BaseCloudSolrClient.java:896)
    [junit4]    > 	at org.apache.solr.client.solrj.impl.BaseCloudSolrClient.request(BaseCloudSolrClient.java:828)
    [junit4]    > 	at org.apache.solr.client.solrj.SolrRequest.process(SolrRequest.java:207)
    [junit4]    > 	at org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:177)
    [junit4]    > 	at org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:138)
    [junit4]    > 	at org.apache.solr.cloud.TestTlogReplica.doReplaceLeader(TestTlogReplica.java:410)
    [junit4]    > 	at org.apache.solr.cloud.TestTlogReplica.testRemoveLeader(TestTlogReplica.java:320)

**Comment by janhoy** [comment]
I think this is due to the change of mapping for dispatch filter. Can you have a look?

**Comment by uschindler** [comment]
Does this fail reproducible. From the error message this looks common, I also see it on Jenkins. If that url won't accept POST, more than that would fail.

It could be something else related to the 404. You removed the servlet, but it looks like the dispatch filter is installed a little bit delayed. Maybe the test is accessing to early and instead of 404 and retry it fails because method is incorrect?

What happens if the 404 servlet stays alive?

Uwe

**Comment by uschindler** [comment]
I have the feeling this issues comes from the fact that without the default 404 servlet, Jetty returns some arbitrary error (like invalid method) when something is not found.

Maybe we should revive the hard 404 if the filter does not handle some path? I am not fully sure, but I can look into this later!

We should at least make sure that SolrDispatchFilter returns 404 not found when path is invalid. In addition during bootup time of a new server we should make sure that a valid error is returned.

File: solr/core/src/java/org/apache/solr/client/solrj/embedded/JettySolrRunner.java [hunk 1]
```diff
@@ -381,7 +381,8 @@ public void lifeCycleStarted(LifeCycle arg0) {
         dispatchFilter = root.getServletHandler().newFilterHolder(Source.EMBEDDED);
         dispatchFilter.setHeldClass(SolrDispatchFilter.class);
         dispatchFilter.setInitParameter("excludePatterns", excludePatterns);
-        root.addFilter(dispatchFilter, "*", EnumSet.of(DispatcherType.REQUEST));
+        // Map dispatchFilter in same path as in web.xml
+        root.addFilter(dispatchFilter, "/*", EnumSet.of(DispatcherType.REQUEST));
```
[the remark this answers is not in the document]
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/client/solrj/embedded/JettySolrRunner.java]: Done

**Comment by janhoy** [comment]
I added back the 404 servlet, but if using `/*` path then we get back to the old state where pathInfo "takes over for" servletPath.

What happens in this test is that we remote a replica then attempts to add another document, which should then succeed towards the leader. Have not been able to trace the call to the exact spot it crashes. But the fact that this works with the 404 fallback could of course play a role here.

**Comment by uschindler** [comment]
Ok, I know.

For the default servlet, the path has to be just "/". This is how it's implicitly configured for an empty webapp by jetty (it uses a special class DefaultServlet for this).

The embedded jetty has no webapp only a context, so there is no default servlet.

**Comment by janhoy** [comment]
Added back the 404 servlet but mounted on "/", which seems to have done the job:

    BUILD SUCCESSFUL
    Total time: 48 minutes 17 seconds

File: solr/core/src/java/org/apache/solr/servlet/ServletUtils.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public abstract class ServletUtils {
+  private ServletUtils() { /* only static methods in this class */ }
+
+  /**
+   * Use this to get the full path after context path "/solr", which is a combination of
+   * servletPath and pathInfo.
+   * @param request the HttpServletRequest object
+   * @return String with path starting with a "/", or empty string if no path
+   */
+  public static String getPathAfterContext(HttpServletRequest request) {
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/servlet/ServletUtils.java]: If you have better suggestions for method name here, then bring it on!

    getPathExceptContext
    getPathWihoutContext
    getServletPathWithInfo

**Pull request merged**